OPC-646 Immersive Classrooom integration stable 11

karendolan opened the pull request:
This is the IC-OC-Integration branch migrated to oc-master vs OC v11. 

It includes CORS support for  requests from *.harvard.edu addresses to enable the Porta Cookie to be sent in the Video Comments request from Immersive Classroom servers..  

The pull from https://github.com/harvard-dce/mh-opsworks-recipes/pull/259 was cherry-picked to f/OPC-646-IC-Integration-stable-11.

karendolan committed:
OPC-621 support for dce_porta cookies on GET from *.harvard.edu domains
OPC-621 moved if origin to the location section of server as required by nginx
Immersive Classroom Chat CORS - only ONE if statement can set headers, and existing ones need to be reset. Also, set defaults variables to prevent duplicate headers
IC-CORS peer reivew, reduce scope to *.dcex.harvard.edu vs *.harvard.edu

OPC-647 Add immersive url to custom.properties.
CHANGELOG comment for immersive_classroom_url reference in OC admin

Always return headers regardless of return code (to help debugging)
OPC-646 code review updates: generalize CORs domain param, and patch default IC url

karendolan commented:
@rute-santos Even though it does no harm, should that IC reference from OC admin be removed from this pull? It is only needed for the OC re-trim call to IC feature which has not gone through QA.

lbjay commented on templates/default/engage-nginx-proxy-conf.erb:
```diff
@@ -107,9 +107,38 @@ server {
+  set $cors_vary "";
+  set $cors_expose_headers "";
+
+  if ($http_origin ~* (https?://.*\.dcex.harvard.edu(:[0-9]+)?)) {
```
did you look at alternatives to baking the hostname pattern into the code? We've managed to avoid this so far.

karendolan replied:
Updated this to be generic. It now extracts the domain from the engage public host name. 
Also patched an issue with unescaped dots. this is what the rendered update looks like when deployed to dev Engage node.
`if ($http_origin ~* (https?://.*\.dcex\.harvard\.edu(:[0-9]+)?)) {`

lbjay commented on templates/default/custom.properties.erb:
```diff
@@ -388,3 +388,6 @@ edu.harvard.dce.production.email=<%= @production_management_email %>
 edu.harvard.dce.cas.service=<%= @cas_service %>
 
+# Immersive classroom base url. Used to push a notification when a video is replaced by its trimmed version.
+edu.harvard.dce.immersive.classroom.url=<%= @immersive_classroom_url %>
```
Will it cause problems for this to get the string 'None' when `immersive_classroom_url` isn't set?

karendolan replied:
Hi Jay, good catch. This param is part of Rute's notify IC feature. Her OC feature code snippet below, shows that this is expecting a string or empty string. So, I'll change the default to this from 'None' to '' empty string.

```
+    // Skip if immersive classroom url was not set. This is OK for a dev cluster.
+    if (StringUtils.isEmpty(immersiveBaseUrl)) {
+      return createResult(Action.SKIP);
+    }
```

karendolan replied:
Changed to "" so the OC code will skip default param when not set in a cluster config.

rute-santos commented:
@karendolan I have no strong feelings. The ic url can stay as it causes no harm or be removed and added later.

karendolan commented:
@lbjay , I updated the pull from your 2 excellent comments.
@rute-santos , I left in the param for the notify IC feature, since it does no harm. I changed the default to "" to match what the OC code expects for default.

lbjay commented:
Looks great. Thanks for the reminder!
